analytics_data_api/v0/models.py

```diff
@@ -220,3 +223,64 @@ class Meta:
     def get_course_user(cls, course_id, username):
         return cls.search().query('term', course_id=course_id).query(
             'term', username=username).execute()
+
+
+class ModuleEngagement(models.Model):
+    """User interactions with entities within the courseware."""
+
+    course_id = models.CharField(db_index=True, max_length=255)
+    username = models.CharField(max_length=255)
+    date = models.DateTimeField()
+    # This will be one of "problem", "video" or "forum"
+    entity_type = models.CharField(max_length=255)
+    # For problems this will be the usage key, for videos it will be the html encoded module ID,
+    # for forums it will be the commentable_id
+    entity_id = models.CharField(max_length=255)
+    # A description of what interaction occurred.
+    event = models.CharField(max_length=255)
+    # The number of times the user interacted with this entity in this way on this day.
+    count = models.IntegerField()
+
+    class Meta(object):
+        db_table = 'module_engagement'
+
+
+class ModuleEngagementTimelineManager(models.Manager):
+    """
+    Modifies the ModuleEngagement queryset to aggregate engagement data for
+    the learner engagement timeline.
+    """
+    def get_timelines(self, course_id, username):
+        queryset = ModuleEngagement.objects.all().filter(course_id=course_id, username=username)\
+            .values('date', 'entity_type', 'event')\
+            .annotate(count=Sum('count'))\
+            .order_by('date')
+
+        timelines = []
+
+        for key, group in groupby(queryset, lambda x: (x['date'])):
```

Comment:
Performance conern:

We have (up to) `n` days and `m` modules that a student interacts with on any given day. Theoretically, the queryset is as large as `O(n * m)`, and I assume that `itertools.groupby` could be an `O(n^2)` operation, so we may have `O((n * m)^2` here.

@mulby and I chatted about this, and he's optimistic that in practice, `m` will be very small, so he's not terribly worried. Regardless, we explored a few ways to improve the performance of this code:

* Push this group by logic into SQL. We might be able to do this by making the above query a sub-query to a SELECT ... GROUP_BY clause. We're not sure how the `entity_type`/`event` joining logic would work, though.
* If we can't push the logic into SQL, find out if we can stream results out of SQL (do querysets work that way?) into memory, and manage a `dict` which maps date strings to engagement counts, thereby reducing the space complexity. Instead of the quadratic group_by, you can just iterate through the results one at a time, either adding a new date, or updating an existing one. For example, the dict would have the form:
```python
{
    '01-01-01': {
        'problems_attempted': 0,
        ....
    },
    ...
}
```